### PR TITLE
Extended dshot telemetry

### PR DESCRIPTION
**Pull request by damosvil**
Created a new pr for edt because for an unknown reason the old one doesn't update new commits.
This pr has the old pr findings solved.

### CODE REVIEW

File: Bluejay.asm [hunk 1]
```diff
@@ -1662,95 +1662,186 @@ set_pwm_limit_high_rpm_store:
...
 	clr	C
-	mov	A, Adc_Conversion_Cnt		; Is conversion count equal to temp rate?
-	subb	A, #8
-	jc	temp_increase_pwm_limit		; No - increase pwm limit
+	subb	A, Temp_Prot_Limit						; Is temperature below first limit?
+	jc	scheduler_1024ms_switch_case				; Yes - Jump to next scheduler
+
+	mov	Temp_Pwm_Level_Setpoint, #200				; No - update pwm limit (about 80%)
+
+	subb	A, #(TEMP_LIMIT_STEP / 2)				; Is temperature below second limit
```
Comment (stylesuxx) [inline Bluejay.asm:1708]: Probably a stupid question: division by two is the same as right shifting the value once - or will this part be assembled into something different?

**Comment by stylesuxx** [comment]
In comparison to the "old" PR, this one includes:

* Extended Dshot Telemetry
* Improved handling of temperature

Looks very clean to me - thank you @damosvil

I will adapt the Arduino DSHOT sender accordingly to enable and read EDT.

File: Bluejay.asm (hunk 1 above)
Comment (damosvil) [reply Bluejay.asm:1708]: That division is handled by the Keil assemby preprocessor

Comment (stylesuxx) [reply Bluejay.asm:1708]: Ah yeah, makes sense. TEMP_LIMIT_STEP does not change during execution...

**Review by stylesuxx — APPROVED** [review]

**Comment by howels** [comment]
Tested on a G-H-30 target, working great.

**Comment by howels** [comment]
> > Tested on a G-H-30 target, working great.
> 
> Thank you for testing. Did you test this PR or the "old" one?
> 
> > Only problem is locating an ESC with per-motor current/voltage sensors.
> 
> Yeah, I guess this might change in the future once the PR has been merged and we have something to ask manufacturers to implement...

I tested hex sent from @damosvil, unfortunately because this PR doesn't have files attached. If building was more straightforward I could run a local build but I haven't got round to licensing the compiler. Perhaps GitHub workflows could build PRs in future?

**Comment by stylesuxx** [comment]
> I tested hex sent from @damosvil, unfortunately this PR doesn't have files attached. If building was more straightforward I could run a local build but I haven't got round to licensing the compiler. Perhaps GitHub workflows could build PRs in future?

Yes, this is planned - we just did not yet decide on how we solve the licence situation. Building is actually pretty straight forward once you have the stupid licence. You can ping me if you need assistance.

**Comment by damosvil** [comment]
Sent to draft again to create a patch to reset extended dshot activated flag after disarming the kwad. Fc sends it again when arming

File: Bluejay.asm [hunk 2]
```diff
@@ -2912,17 +3033,30 @@ dshot_tlm_create_packet:
...
 	mov	Tlm_Data_H, #0Fh
 	mov	Tlm_Data_L, #0FFh
+	sjmp dshot_tlm_12bit_encoded
+
+dshot_tlm_ext_coded:
+	; Move extended telemetry data to telemetry data to send
+	mov Tlm_Data_L, Ext_Telemetry_L
+	mov Tlm_Data_H, Ext_Telemetry_H
+	; Clear extended telemetry data
+	mov Ext_Telemetry_H, #0
```
Comment (saidinesh5) [inline Bluejay.asm:3060]: Only clearing Ext_Telemetry_H and not _L because L will always be written to?

File: Bluejay.asm [hunk 3]
```diff
@@ -3924,6 +4060,7 @@ beep_delay_set:
 wait_for_start_no_beep:
 	jb	Flag_Telemetry_Pending, wait_for_start_check_rcp
 	call	dshot_tlm_create_packet		; Create telemetry packet (0 rpm)
+	call 	scheduler_run
```
Comment (saidinesh5) [inline Bluejay.asm:4063]: Should this happen after the beep routine is done? because during beep routines we want no interrupts.. and if scheduler starts a timer which causes an interrupt, then the registers used by beep routine may get overwritten

File: Bluejay.asm [hunk 4]
```diff
@@ -4238,6 +4366,7 @@ exit_run_mode_on_timeout:
 
 exit_run_mode:
 	clr	IE_EA					; Disable all interrupts
+	clr Flag_Ext_Tele			; Clear extended DSHOT telemetry flag
```
Comment (saidinesh5) [inline Bluejay.asm:4369]: I don't care much about what formatting we use as long as it is consistent. Some places it is 1 space, some places it is 4 spaces. Some places tabs. So i say we use  some git commit hooks to run a formatter before commit.

Note, this is NOT a suggestion for this pull request. Has to be done in a separate pull request - when done.

File: Bluejay.asm [hunk 5]
```diff
@@ -3959,26 +4096,16 @@ motor_start:
...
-	mov	Current_Average_Temp, ADC0L	; Read initial temperature
-	mov	A, ADC0H
-	jnz	($+5)					; Is reading below 256?
-	mov	Current_Average_Temp, #0		; Yes - set average temperature value to zero
-
-	mov	Adc_Conversion_Cnt, #8		; Make sure a temp reading is done
-	call	check_temp_and_limit_power
-	mov	Adc_Conversion_Cnt, #8		; Make sure a temp reading is done next time
+	mov Ext_Telemetry_H, #0				; Clear extended telemetry data
+	mov	Temp_Pwm_Level_Setpoint, #255	; Initialize temperature pwm level setpoint
```
Comment (saidinesh5) [inline Bluejay.asm:4100]: This mov is useless because we are writing Pwm_Limit_Beg into Temp_Pwm_Level_Setpoint right below this.

File: Bluejay.asm (hunk 4 above)
Comment (stylesuxx) [reply Bluejay.asm:4369]: Yes, totally, I already added this to the [cleanup task](#13).

File: Bluejay.asm (hunk 2 above)
Comment (damosvil) [reply Bluejay.asm:3060]: Ext_Telemetry_H has double purpose, on one hand it carries the high part (ext. telemetry type) of the extended telemetry value. On the other hand when extended telemetry is checked to be sent this value is compared to 0, and if different, extended telemetry is sent instead of erpm coded value. For that reason setting it to 0 after sending extended telemetry disables sending it again. Ext_Telemetry_L is left as vestigial value, and nothing is done with that variable.

File: Bluejay.asm (hunk 3 above)
Comment (damosvil) [reply Bluejay.asm:4063]: this call is inserted here to allow sending extended telemetry while waiting for the initial rc pulse value, for example in Betaflight motors tab, so you don't need to spin the motors to start seeing temperature values. This was done to replicate Blheli_32 behaviour.

File: Bluejay.asm (hunk 5 above)
Comment (damosvil) [reply Bluejay.asm]: Fixed

File: Bluejay.asm (hunk 3 above)
Comment (stylesuxx) [reply Bluejay.asm:4063]: @damosvil I am not sure if this answers the question - will there be interrupts because of that while the beep routine is running?

Comment (damosvil) [reply Bluejay.asm:4063]: Np, there are no interrupts resulting of scheduler_run call

**Comment by howels** [comment]
If the workflow is able to build this PR then I'd be happy to test it out and put a few packs through.

**Comment by stylesuxx** [comment]
> If the workflow is able to build this PR then I'd be happy to test it out and put a few packs through.

Good point, the runner is failing with memory overlaps from the assembler.

@damosvil is the latest version you pushed building for you locally?

The automated build says the following:
```
fatal: No names found, cannot describe anything.
AX51 : build/A_H_0_24_.OBJ
LX51 : linking build/A_H_0_24_.OBJ to build/A_H_0_24_.OMF
*** WARNING L30: MEMORY SPACE OVERLAP
    SEG:     ?DT?BLUEJAY?22
    FROM:    I:000020H   
    TO:      I:000028H   
--
make: *** [Makefile:99: build/A_H_0_24_.OMF] Error 1
*** ERROR L121: IMPROPER FIXUP
    MODULE:  BUILD/A_H_0_24_.OBJ (BLUEJAY)
    SEGMENT: ?CO?BLUEJAY?20
    OFFSET:  0008[7](https://github.com/bird-sanctuary/bluejay/actions/runs/3169807556/jobs/5161966437#step:6:8)5H
```

**Comment by damosvil** [comment]
Building one target at a time I have no problems. The other day I tried to build all and it also failed

**Comment by stylesuxx** [comment]
@damosvil The issue is true for all 0 dead-time targets. Everything else seems to build fine.

**Comment by damosvil** [comment]
Ok, I will try to have a look asap

**Comment by stylesuxx** [comment]
@howels fresh build, get the hex'es while they are still hot...

@saidinesh5 Have your questions been answered satisfactory? If so I am soooo ready to merge this one ;-)

@damosvil Thanks for the quick fix!

**Review by stylesuxx — APPROVED** [review]

**Comment by saidinesh5** [comment]
@stylesuxx yeah - for me the two major concerns were if the interrupts overwrite some important registers and put us in an undefined state - Especially during startup - when it would lead to a bootloop and bricked ESCs for sure. ( most of the timer 2 interrupt code didn't seem like it. @damosvil clarified that this won't be the case during the startup either. )

Wanted to make sure the builds pass, and you got that to work.

I couldn't find much time to test this on my real hardware yet .. but i am happy with the code. Feel free to merge this. If there are any problems - especially detected with your arduino tester, we fix them before making a release

**Comment by stylesuxx** [comment]
@saidinesh5 could you just resolve your open conversations and approve the PR - only to get a bit structure into it :-D

**Review by saidinesh5 — APPROVED** [review]
